ariadne/src/Ariadne/Wallet/Backend/Tx.hs

```diff
@@ -74,22 +77,55 @@ sendTx ::
     -> InputSelectionPolicy
     -> NonEmpty TxOut
     -> IO TxId
-sendTx acidDb WalletFace {..} CardanoFace {..} walletSelRef printAction pp walletRef _accRefs isp outs = do
+sendTx acidDb WalletFace {..} CardanoFace {..} walletSelRef printAction pp walletRef accRefs isp outs = do
     let Nat runCardanoMode = cardanoRunCardanoMode
     walletDb <- query acidDb Snapshot
     let wallets = walletDb ^. dbHdWallets
-    walletRootId <- resolveWalletRef walletSelRef walletRef walletDb
-    runCardanoMode $ sendTxDo wallets walletRootId =<< cardanoGetDiffusion
+    (walletRootId, walletAccounts) <-
+        resolveWalletRefThenRead walletSelRef walletRef
+        walletDb readAccountsByRootId
+    accountIds <- resolveAccountRefs walletRootId walletAccounts
+    let filteredAccounts :: IxSet HdAccount
+        filteredAccounts = maybe identity filterAccounts accountIds walletAccounts
+
+        filterAccounts :: NonEmpty HdAccountId -> IxSet HdAccount -> IxSet HdAccount
+        filterAccounts ids accounts = accounts @+ toList ids
+
+    runCardanoMode $
+        sendTxDo wallets walletRootId filteredAccounts =<< cardanoGetDiffusion
   where
-    sendTxDo :: HdWallets -> HdRootId -> Diffusion CardanoMode -> CardanoMode TxId
-    sendTxDo wallets walletRootId diffusion = do
-        let
-            walletAccounts :: IxSet HdAccount
-            walletAccounts = fromRight
-                -- Resolved walletRootId always exists.
-                (error "SendTx: Unknown walletRootId")
-                (readAccountsByRootId walletRootId wallets)
-
+    -- Returns list of accounts which can be used.
+    -- 'Nothing' means all accounts can be used.
+    resolveAccountRefs ::
+           HdRootId
+        -> IxSet HdAccount
+        -> IO (Maybe (NonEmpty HdAccountId))
+    resolveAccountRefs rootId accountsSet =
+        case nonEmpty accRefs of
+            Nothing -> do
+                readIORef walletSelRef >>= \case
+                    Just (WSAccount selectedAccId)
+                        | selectedAccId ^. hdAccountIdParent == rootId ->
```

Comment:
As far as I see, this `rootId` comes from `resolveWalletRef`, but here you are checking that `walletSelRef` contains a `HdAccountId` referring to the same `rootId` that has already been returned by `resolveWalletRef`. I haven't read what `resolveWalletRef` does in detail or where `walletSelRef` comes from, but are you sure that this extra check is needed here?

Reply:
See the documentation of this function.
```
If list of accounts is not empty, only those accounts
may be used as inputs.  If this list is empty and an account from
the input wallet is selected, this account will be used as input.
Otherwise inputs will be selected from all accounts in the wallet.
```

This case is when an account from the input wallet is selected.

Reply:
Ok, LGTM.